src/watchman_client.js

```diff
@@ -0,0 +1,376 @@
+'use strict';
+
+var watchman = require('fb-watchman');
+
+/**
+ * Constants
+ */
+
+/**
+ * Singleton that provides a public API for a single connection to a watchman instance for 'sane'.
+ * It tries to abstract/remove as much of the boilerplate processing as necessary
+ * from WatchmanWatchers that use it. In particular, they have no idea whether
+ * we're using 'watch-project' or 'watch', what the 'project root' is when
+ * we internally use watch-project, whether a connection has been lost
+ * and reestablished, etc. Also switched to doing things with promises and known-name
+ * methods in WatchmanWatcher, so as much information as possible can be kept in 
+ * the WatchmanClient, ultimately making this the only object listening directly
+ * to watchman.Client, then forwarding appropriately (via the known-name methods) to 
+ * the relevant WatchmanWatcher(s).
+ *
+ * @class WatchmanClient
+ * @param String dir
+ * @param {Object} opts
+ * @public
+ */
+
+function WatchmanClient() {
+
+  // define/clear some local state. The properties will be initialized
+  // in getClient(). This is also called again in _onEnd when
+  // trying to reestablish connection to watchman.Client.
+  this._clearLocalVars(); 
+
+  this._clientListeners = null;  // direct listeners from here to watchman.Client.
+
+  // we may want to call 'getClient()' and return a promise instead later,
+  // but for now we'll assume the WatchmanWatchers will not use wildmatch
+  // until they've done a getClient to retrieve the instance, so the
+  // capabilityCheck call is done.
+  Object.defineProperty(this, 'wildmatch', {
```

Comment:
lets define this on the prototype, rather then on each instance.

```
Object.defineProperty(WatchmanClient.prototype, 'wildmatch', {
  get() { return this._wildmatch; }
});
```

That will then mimic what happens with ES6 classes `get wildmatch() { ` syntax.

Reply:
I wasn't actually sure you could define a getter on the prototype for an instance var - that's why I did it this way. Happy to switch, though, as I agree with the comment.